### PR TITLE
Add optimizations for HyMD

### PR DESCRIPTION
**Pull request by MichaelS239**

### CODE REVIEW

File: src/core/algorithms/md/hymd/lattice_traverser.h [hunk 1]
```diff
@@ -12,7 +12,14 @@ namespace algos::hymd {
...
 class LatticeTraverser {
 private:
+    struct LatticeStatistics {
+        std::size_t all_mds_num = 0;
+        std::size_t invalidated_mds_num = 0;
+        static constexpr double kRatioBound = 0.01;
+    };
+
     Recommendations recommendations_;
+    LatticeStatistics lattice_statistics_;
```
Comment (wildsor) [inline src/core/algorithms/md/hymd/lattice_traverser.h]: These statistics should only be collected for the current level, not for the whole lattice traversal

File: src/core/algorithms/md/hymd/validator.cpp [hunk 2]
```diff
@@ -431,7 +431,7 @@ void Validator::MakeWorkingAndRecs(lattice::ValidationInfo const& info,
     std::vector<ColumnClassifierValueId> const removed_ccv_ids =
             lattice_->RemoveExisting(lhs, indices);
     std::vector<ColumnClassifierValueId> const interestingness_ccv_ids =
-            lattice_->GetInterestingnessCCVIds(lhs, indices);
+            lattice_->GetInterestingnessCCVIds(lhs, indices, removed_ccv_ids);
```
Comment (wildsor) [inline src/core/algorithms/md/hymd/validator.cpp:436]: I just figured out that you can decrement every element of `removed_ccv_ids`, pass that, then increment it back. Rename the `Raise...` and `Get...` methods' parameter from `removed_ccv_ids` to `ccv_id_bounds` too.

File: src/core/algorithms/md/hymd/utility/size_t_vector_hash.h [hunk 3]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <cstddef>
```
Comment (wildsor) [inline src/core/algorithms/md/hymd/utility/size_t_vector_hash.h:3]: I think this is only used to hash vectors of `ColumnClassifierValueId`, you should change it to that everywhere (includes, hash definition, filename)

File: src/core/algorithms/md/hymd/similarity_data.cpp [hunk 4]
```diff
@@ -15,8 +15,11 @@ SimilarityData SimilarityData::CreateFrom(indexes::RecordsInfo* const records_in
     column_matches_info.reserve(col_match_number);
     std::vector<std::vector<model::Index>> column_matches_lhs_ids;
     column_matches_lhs_ids.reserve(col_match_number);
+    std::vector<std::pair<model::Index, std::size_t>> lhs_size_to_index;
```
Comment (wildsor) [inline src/core/algorithms/md/hymd/similarity_data.cpp]:
```suggestion
    std::vector<std::pair<std::size_t, model::Index>> lhs_size_to_index;
```

**Review by wildsor — CHANGES_REQUESTED** [review]

File: src/core/algorithms/md/hymd/utility/size_t_vector_hash.h (hunk 3 above)
Comment (MichaelS239) [reply src/core/algorithms/md/hymd/utility/size_t_vector_hash.h:3]: The hash is also used for `Validator::MultiCardPairProvider::GroupMap`

File: src/core/algorithms/md/hymd/lattice_traverser.cpp [hunk 5]
```diff
@@ -6,12 +6,16 @@ namespace algos::hymd {
 
 bool LatticeTraverser::TraverseLattice(bool const traverse_all) {
     using model::Index;
+    LatticeStatistics lattice_statistics;
     while (level_getter_->AreLevelsLeft()) {
```
Comment (wildsor) [inline src/core/algorithms/md/hymd/lattice_traverser.cpp:10]:
```suggestion
    while (level_getter_->AreLevelsLeft()) {
        LatticeStatistics lattice_statistics;
```
Maybe your version is also okay, but this is what I'm looking for.

File: src/core/algorithms/md/hymd/validator.cpp [hunk 6]
```diff
@@ -428,10 +428,15 @@ void Validator::MakeWorkingAndRecs(lattice::ValidationInfo const& info,
     std::size_t const working_size = indices.size();
     working.reserve(working_size);
     recommendations.reserve(working_size);
-    std::vector<ColumnClassifierValueId> const removed_ccv_ids =
-            lattice_->RemoveExisting(lhs, indices);
+    std::vector<ColumnClassifierValueId> removed_ccv_ids = lattice_->RemoveExisting(lhs, indices);
+    for (std::size_t i = 0; i < removed_ccv_ids.size(); ++i) {
+        --removed_ccv_ids[i];
+    }
```
Comment (wildsor) [inline src/core/algorithms/md/hymd/validator.cpp:434]:
```suggestion
    for (ColumnClassifierValueId& ccv_id : removed_ccv_ids) {
        --ccv_id;
    }
```

File: src/core/algorithms/md/hymd/validator.cpp [hunk 7]
```diff
@@ -428,10 +428,15 @@ void Validator::MakeWorkingAndRecs(lattice::ValidationInfo const& info,
...
     std::vector<ColumnClassifierValueId> const interestingness_ccv_ids =
-            lattice_->GetInterestingnessCCVIds(lhs, indices);
+            lattice_->GetInterestingnessCCVIds(lhs, indices, removed_ccv_ids);
+    for (std::size_t i = 0; i < removed_ccv_ids.size(); ++i) {
+        ++removed_ccv_ids[i];
+    }
```
Comment (wildsor) [inline src/core/algorithms/md/hymd/validator.cpp:439]:
```suggestion
    for (ColumnClassifierValueId& ccv_id : removed_ccv_ids) {
        ++ccv_id;
    }
```

**Review by wildsor — APPROVED** [review]
Good job, I'll fix those comments myself

**Pull request merged**